### PR TITLE
NIAD-1182: Global Logging

### PR DESCRIPTION
**Pull request by ElliotB-96**
- Added MDC handling
- updated logback to better meet specified format

### CODE REVIEW

**Review by z-uddin — APPROVED** [review]
From looking at the pipeline logs for the tests can observe Ssp-TraceId is logged when Ssp-TraceId is present. And RequestId being logged when a request is made to the GPCC adaptor in the test logs.

**Review by mattd-kainos — CHANGES_REQUESTED** [review]
Doesn't meet AC 1 - nothing is logged for requests that are not routed

File: service/src/main/java/uk/nhs/adaptors/gpc/consumer/filters/ReactorMdcSupport.java [hunk 1]
```diff
@@ -0,0 +1,95 @@
...
+
+    private static final String MDC_HOOK_KEY = "MDC";
+    private static final Function<? super Publisher<Object>, ? extends Publisher<Object>> LIFTER =
+        Operators.liftPublisher(
+            publisher -> {
+                if (MDC.getCopyOfContextMap() == null) {
+                    return false;
+                }
+                return !(publisher instanceof Fuseable.ScalarCallable);
+            },
```
Comment (andrewwa-kainos) [inline service/src/main/java/uk/nhs/adaptors/gpc/consumer/filters/ReactorMdcSupport.java:32]: Possible alternative
```suggestion
            publisher -> MDC.getCopyOfContextMap() != null
                && !(publisher instanceof Fuseable.ScalarCallable),
```

File: service/src/main/java/uk/nhs/adaptors/gpc/consumer/filters/ReactorMdcSupport.java [hunk 2]
```diff
@@ -0,0 +1,95 @@
...
+            ((publisher, coreSubscriber) -> new MdcPropagatingSubscriber<>(coreSubscriber))
```
Comment (andrewwa-kainos) [inline service/src/main/java/uk/nhs/adaptors/gpc/consumer/filters/ReactorMdcSupport.java:33]: Minor point: the wrapping parentheses are superfluous, but harmless
```suggestion
            (publisher, coreSubscriber) -> new MdcPropagatingSubscriber<>(coreSubscriber)
```

File: service/src/main/java/uk/nhs/adaptors/gpc/consumer/filters/ReactorMdcSupport.java [hunk 3]
```diff
@@ -0,0 +1,95 @@
...
+        MdcPropagatingSubscriber(CoreSubscriber<T> delegate) {
+            this.delegate = delegate;
+            Context currentContext = this.delegate.currentContext();
+            Context context;
+            if (currentContext.hasKey(MDC_CONTEXT_KEY)) {
+                context = currentContext;
+            } else {
+                Map<String, String> map = new HashMap<>(MDC.getCopyOfContextMap());
+                context = currentContext.put(MDC_CONTEXT_KEY, map);
+            }
```
Comment (andrewwa-kainos) [inline service/src/main/java/uk/nhs/adaptors/gpc/consumer/filters/ReactorMdcSupport.java:61]: If you're comfortable with the ternary conditional operator
```suggestion
            Context context = currentContext.hasKey(MDC_CONTEXT_KEY)
                ? currentContext
                : currentContext.put(MDC_CONTEXT_KEY, new HashMap<>(MDC.getCopyOfContextMap()));
```

File: service/src/main/java/uk/nhs/adaptors/gpc/consumer/filters/ReactorMdcSupport.java [hunk 4]
```diff
@@ -0,0 +1,95 @@
...
+import org.springframework.context.annotation.Configuration;
+
+import reactor.core.CoreSubscriber;
+import reactor.core.Fuseable;
+import reactor.core.publisher.Hooks;
+import reactor.core.publisher.Operators;
+import reactor.util.context.Context;
+
+@Configuration
+public class ReactorMdcSupport {
```
Comment (andrewwa-kainos) [inline service/src/main/java/uk/nhs/adaptors/gpc/consumer/filters/ReactorMdcSupport.java:22]: This class seems to be a decorator - so could we call it `ReactorMdcSupportDecorator` to make that clear?

File: service/src/main/java/uk/nhs/adaptors/gpc/consumer/utils/MdcUtil.java [hunk 5]
```diff
@@ -0,0 +1,19 @@
...
+import org.springframework.web.server.ServerWebExchange;
+
+public class MdcUtil {
+    private static final String REQUEST_ID = "RequestId";
+    private static final String TRACE_ID = "Ssp-TraceID";
+
+    public static void applyHeadersToMdc(ServerWebExchange exchange) {
+        MDC.put(REQUEST_ID, exchange.getLogPrefix());
+        HttpHeaders headers = exchange.getRequest().getHeaders();
+        var traceId = headers.get(TRACE_ID);
```
Comment (andrewwa-kainos) [inline service/src/main/java/uk/nhs/adaptors/gpc/consumer/utils/MdcUtil.java:14]: For anyone unfamiliar with spring it won't be immediately clear what type `traceId` is so in this case I'd advise against `var`.
```suggestion
        List<String> traceId = headers.get(TRACE_ID);
```

File: service/src/main/java/uk/nhs/adaptors/gpc/consumer/utils/MdcUtil.java [hunk 6]
```diff
@@ -0,0 +1,19 @@
...
+        if (traceId != null && traceId.size() > 0) {
```
Comment (andrewwa-kainos) [inline service/src/main/java/uk/nhs/adaptors/gpc/consumer/utils/MdcUtil.java:15]: Lists have a method for this
```suggestion
        if (traceId != null && !traceId.isEmpty()) {
```

**Review by andrewwa-kainos — COMMENTED** [review]
Some observations - all minor; nothing that I'd insist on

**Review by mattd-kainos — APPROVED** [review]
looks great, logging works for both routed and non-routed paths including identifiers

**Review by z-uddin — APPROVED** [review]

**Pull request merged**